// pg-structure/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Trigger = exports.WindowFunction = exports.FunctionArgument = exports.AggregateFunction = exports.Procedure = exports.NormalFunction = exports.Func = exports.MultiRangeType = exports.RangeType = exports.PseudoType = exports.EnumType = exports.Domain = exports.CompositeType = exports.BaseType = exports.O2MRelation = exports.M2ORelation = exports.M2MRelation = exports.View = exports.Table = exports.Sequence = exports.MaterializedView = exports.UniqueConstraint = exports.PrimaryKey = exports.ForeignKey = exports.ExclusionConstraint = exports.CheckConstraint = exports.Type = exports.Relation = exports.Entity = exports.DbObject = exports.Constraint = exports.Schema = exports.Index = exports.Db = exports.Column = exports.deserialize = void 0;
const main_1 = require("./main");
Object.defineProperty(exports, "deserialize", { enumerable: true, get: function () { return main_1.deserialize; } });
const column_1 = __importDefault(require("./pg-structure/column"));
exports.Column = column_1.default;
const db_1 = __importDefault(require("./pg-structure/db"));
exports.Db = db_1.default;
const index_1 = __importDefault(require("./pg-structure/index"));
exports.Index = index_1.default;
const schema_1 = __importDefault(require("./pg-structure/schema"));
exports.Schema = schema_1.default;
const constraint_1 = __importDefault(require("./pg-structure/base/constraint"));
exports.Constraint = constraint_1.default;
const db_object_1 = __importDefault(require("./pg-structure/base/db-object"));
exports.DbObject = db_object_1.default;
const entity_1 = __importDefault(require("./pg-structure/base/entity"));
exports.Entity = entity_1.default;
const relation_1 = __importDefault(require("./pg-structure/base/relation"));
exports.Relation = relation_1.default;
const type_1 = __importDefault(require("./pg-structure/base/type"));
exports.Type = type_1.default;
const check_constraint_1 = __importDefault(require("./pg-structure/constraint/check-constraint"));
exports.CheckConstraint = check_constraint_1.default;
const exclusion_constraint_1 = __importDefault(require("./pg-structure/constraint/exclusion-constraint"));
exports.ExclusionConstraint = exclusion_constraint_1.default;
const foreign_key_1 = __importDefault(require("./pg-structure/constraint/foreign-key"));
exports.ForeignKey = foreign_key_1.default;
const primary_key_1 = __importDefault(require("./pg-structure/constraint/primary-key"));
exports.PrimaryKey = primary_key_1.default;
const unique_constraint_1 = __importDefault(require("./pg-structure/constraint/unique-constraint"));
exports.UniqueConstraint = unique_constraint_1.default;
const materialized_view_1 = __importDefault(require("./pg-structure/entity/materialized-view"));
exports.MaterializedView = materialized_view_1.default;
const sequence_1 = __importDefault(require("./pg-structure/entity/sequence"));
exports.Sequence = sequence_1.default;
const table_1 = __importDefault(require("./pg-structure/entity/table"));
exports.Table = table_1.default;
const view_1 = __importDefault(require("./pg-structure/entity/view"));
exports.View = view_1.default;
const m2m_relation_1 = __importDefault(require("./pg-structure/relation/m2m-relation"));
exports.M2MRelation = m2m_relation_1.default;
const m2o_relation_1 = __importDefault(require("./pg-structure/relation/m2o-relation"));
exports.M2ORelation = m2o_relation_1.default;
const o2m_relation_1 = __importDefault(require("./pg-structure/relation/o2m-relation"));
exports.O2MRelation = o2m_relation_1.default;
const base_type_1 = __importDefault(require("./pg-structure/type/base-type"));
exports.BaseType = base_type_1.default;
const composite_type_1 = __importDefault(require("./pg-structure/type/composite-type"));
exports.CompositeType = composite_type_1.default;
const domain_1 = __importDefault(require("./pg-structure/type/domain"));
exports.Domain = domain_1.default;
const enum_type_1 = __importDefault(require("./pg-structure/type/enum-type"));
exports.EnumType = enum_type_1.default;
const pseudo_type_1 = __importDefault(require("./pg-structure/type/pseudo-type"));
exports.PseudoType = pseudo_type_1.default;
const range_type_1 = __importDefault(require("./pg-structure/type/range-type"));
exports.RangeType = range_type_1.default;
const multi_range_type_1 = __importDefault(require("./pg-structure/type/multi-range-type"));
exports.MultiRangeType = multi_range_type_1.default;
const func_1 = __importDefault(require("./pg-structure/base/func"));
exports.Func = func_1.default;
const normal_function_1 = __importDefault(require("./pg-structure/function/normal-function"));
exports.NormalFunction = normal_function_1.default;
const procedure_1 = __importDefault(require("./pg-structure/function/procedure"));
exports.Procedure = procedure_1.default;
const aggregate_function_1 = __importDefault(require("./pg-structure/function/aggregate-function"));
exports.AggregateFunction = aggregate_function_1.default;
const function_argument_1 = __importDefault(require("./pg-structure/function-argument"));
exports.FunctionArgument = function_argument_1.default;
const window_function_1 = __importDefault(require("./pg-structure/function/window-function"));
exports.WindowFunction = window_function_1.default;
const trigger_1 = __importDefault(require("./pg-structure/trigger"));
exports.Trigger = trigger_1.default;
__exportStar(require("./types/index"), exports);
exports.default = main_1.pgStructure;
